src/protocol/sort/generate_sort_permutation.rs

```diff
@@ -0,0 +1,146 @@
+        modulus_conversion::convert_shares::convert_shares_for_a_bit,
+        sort::bit_permutation::BitPermutation,
+        sort::SortStep::{ApplyInv, BitPermutationStep, ComposeStep, ModulusConversion},
+        IpaProtocolStep::Sort,
+    },
+    secret_sharing::Replicated,
+};
+
+use super::{compose::Compose, secureapplyinv::SecureApplyInv};
+use embed_doc_image::embed_doc_image;
+
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[derive(Debug)]
+pub struct GenerateSortPermutation<'a> {
+    input: &'a [(u64, u64)],
+    num_bits: u8,
+}
+
+impl<'a> GenerateSortPermutation<'a> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [(u64, u64)], num_bits: u8) -> GenerateSortPermutation {
+        Self { input, num_bits }
+    }
+
+    #[allow(dead_code)]
+    #[embed_doc_image("semi_honest_sort", "images/sort/semi-honest-sort.png")]
+    /// This protocol generates permutation of a stable sort for the given shares of inputs.
+    /// ![Generate sort permutation steps][semi_honest_sort]
+    /// Steps
+    /// For each bit of input share, following steps are executed
+    /// 1. For the current bit, get replicated shares in Field using modulus conversion
+    /// 2. 1st bit onwards, sort ith bit based on i-1th bits by applying i-1th composition on ith bit
+    /// 3  Sort ith bit by computing bit permutation
+    /// 4. 1st bit onwards, compute ith composition by composing i-1th composition on ith permutation
+    /// In the end, n-1th composition is returned. This is the permutation which sorts the inputs
+    pub async fn execute<F: Field>(
+        &self,
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mut i_minus_1_compose = Vec::with_capacity(self.input.len());
+        for bit_num in 0..self.num_bits {
+            let ctx = ctx.narrow(&Sort(bit_num));
+            let mut bit_value_share = convert_shares_for_a_bit(
+                &ctx.narrow(&ModulusConversion),
+                self.input,
+                self.num_bits,
+                bit_num,
+            )
+            .await?;
+            if bit_num != 0 {
+                SecureApplyInv::execute(
+                    ctx.narrow(&ApplyInv),
+                    &mut bit_value_share,
+                    &mut i_minus_1_compose.clone(),
+                )
+                .await?;
+            }
+            let mut bit_i_permutation = BitPermutation::new(&bit_value_share)
+                .execute(ctx.narrow(&BitPermutationStep))
+                .await?;
+
+            if bit_num != 0 {
+                Compose::new(&mut i_minus_1_compose, &mut bit_i_permutation)
+                    .execute(ctx.narrow(&ComposeStep))
+                    .await?;
+            }
```

Comment:
Originally the paper is written like the way you mention i.e. for bit0 and then a loop from 1..bit_size. I thought this was was easier to understand. Let me restructure and then we can decide. 
Agree on the mutable, checking if it can be done

Reply:
Addressed clear update/modification in #210